File: api/app/mcp_tokens/service.py
from __future__ import annotations

import hashlib
import hmac
import secrets
from datetime import UTC, datetime, timedelta
from typing import Any
from uuid import UUID, uuid4

from app.models import (
    McpTokenCreateRequest,
    McpTokenCreateResponse,
    McpTokenScope,
    McpTokenSummary,
)

from .models import TOKEN_PREFIX, McpTokenAuthContext, McpTokenRecord
from .repository import create_mcp_token, list_mcp_tokens, revoke_mcp_token


def normalize_string_list(values: list[str] | None) -> list[str]:
    """Normalize user-provided list fields (tools/project IDs).

    We trim whitespace, drop blanks, and keep insertion order while deduplicating.
    This keeps token policy matching deterministic across UI/API inputs.
    """
    if not values:
        return []
    result: list[str] = []
    seen: set[str] = set()
    for raw in values:
        value = (raw or "").strip()
        if not value:
            continue
        if value in seen:
            continue
        seen.add(value)
        result.append(value)
    return result


def token_hash(*, token_id: UUID, token_secret: str, pepper: str) -> str:
    payload = f"{token_id.hex}:{token_secret}".encode()
    return hmac.new(pepper.encode(), payload, hashlib.sha256).hexdigest()


def token_secret_hint(token_secret: str) -> str:
    if len(token_secret) <= 10:
        return token_secret
    return f"{token_secret[:6]}...{token_secret[-4:]}"


def build_plaintext_token(*, token_id: UUID, token_secret: str) -> str:
    return f"{TOKEN_PREFIX}_{token_id.hex}_{token_secret}"


def parse_plaintext_token(raw: str) -> tuple[UUID, str]:
    text = (raw or "").strip()
    prefix = f"{TOKEN_PREFIX}_"
    if not text.startswith(prefix):
        raise ValueError("Token prefix mismatch")

    payload = text[len(prefix) :]
    parts = payload.split("_", 1)
    if len(parts) != 2:
        raise ValueError("Token payload format mismatch")

    token_id_hex, token_secret = parts
    if len(token_id_hex) != 32:
        raise ValueError("Token identifier length mismatch")

    try:
        token_id = UUID(hex=token_id_hex)
    except ValueError as exc:
        raise ValueError("Token identifier invalid") from exc

    if not token_secret:
        raise ValueError("Token secret missing")
    return token_id, token_secret


def issue_new_token(
    *,
    token_id: UUID,
    expires_in_days: int,
    pepper: str,
) -> tuple[str, str, str, datetime]:
    """Return plaintext token + persisted hash/hint metadata."""
    expires_at = datetime.now(UTC) + timedelta(days=expires_in_days)
    return issue_token_with_expiry(
        token_id=token_id,
        expires_at=expires_at,
        pepper=pepper,
    )


def issue_token_with_expiry(
    *,
    token_id: UUID,
    expires_at: datetime,
    pepper: str,
) -> tuple[str, str, str, datetime]:
    """Return plaintext token + persisted hash/hint metadata for a fixed expiry."""
    token_secret = secrets.token_urlsafe(32)
    hashed = token_hash(token_id=token_id, token_secret=token_secret, pepper=pepper)
    hint = token_secret_hint(token_secret)
    plaintext = build_plaintext_token(token_id=token_id, token_secret=token_secret)
    return plaintext, hashed, hint, expires_at


def verify_token_secret(
    *, token_id: UUID, token_secret: str, pepper: str, expected_hash: str
) -> bool:
    candidate_hash = token_hash(token_id=token_id, token_secret=token_secret, pepper=pepper)
    return hmac.compare_digest(candidate_hash, expected_hash)


def token_is_active(record: McpTokenRecord, *, now: datetime | None = None) -> bool:
    current = now or datetime.now(UTC)
    if record.revoked_at is not None:
        return False
    return record.expires_at > current


def build_auth_context(*, record: McpTokenRecord) -> McpTokenAuthContext:
    return McpTokenAuthContext(
        token_id=record.token_id,
        owner_user_id=record.owner_user_id,
        scope=record.scope,
        allowed_tools=set(normalize_string_list(record.allowed_tools)),
        allowed_project_ids=set(normalize_string_list(record.allowed_project_ids)),
        expires_at=record.expires_at,
    )


def token_summary_dict(*, record: McpTokenRecord, now: datetime | None = None) -> dict[str, Any]:
    return {
        "token_id": record.token_id,
        "name": record.name,
        "scope": record.scope,
        "allowed_tools": normalize_string_list(record.allowed_tools),
        "allowed_project_ids": normalize_string_list(record.allowed_project_ids),
        "token_secret_hint": record.token_secret_hint,
        "expires_at": record.expires_at,
        "last_used_at": record.last_used_at,
        "revoked_at": record.revoked_at,
        "created_at": record.created_at,
        "is_active": token_is_active(record, now=now),
    }


def create_token_for_owner(
    *,
    owner_user_id: UUID,
    payload: McpTokenCreateRequest,
    pepper: str,
) -> McpTokenCreateResponse:
    normalized_tools = normalize_string_list(payload.allowed_tools)
    normalized_projects = normalize_string_list(payload.allowed_project_ids)

    token_id = uuid4()
    plaintext_token, token_hash_value, token_hint, expires_at = issue_new_token(
        token_id=token_id,
        expires_in_days=payload.expires_in_days,
        pepper=pepper,
    )
    created = create_mcp_token(
        token_id=token_id,
        owner_user_id=owner_user_id,
        name=payload.name.strip(),
        scope=payload.scope.value,
        allowed_tools=normalized_tools,
        allowed_project_ids=normalized_projects,
        token_secret_hash=token_hash_value,
        token_secret_hint=token_hint,
        expires_at=expires_at,
    )
    return McpTokenCreateResponse(
        token_id=created.token_id,
        name=created.name,
        scope=McpTokenScope(created.scope),
        allowed_tools=created.allowed_tools,
        allowed_project_ids=created.allowed_project_ids,
        token_secret_hint=created.token_secret_hint,
        token=plaintext_token,
        expires_at=created.expires_at,
        created_at=created.created_at,
    )


def list_token_summaries(
    *,
    owner_user_id: UUID,
    limit: int = 200,
    offset: int = 0,
    now: datetime | None = None,
) -> list[McpTokenSummary]:
    records = list_mcp_tokens(owner_user_id=owner_user_id, limit=limit, offset=offset)
    current = now or datetime.now(UTC)
    return [McpTokenSummary(**token_summary_dict(record=item, now=current)) for item in records]


def revoke_token_for_owner(
    *,
    token_id: UUID,
    owner_user_id: UUID,
    now: datetime | None = None,
) -> McpTokenSummary | None:
    revoked = revoke_mcp_token(token_id=token_id, owner_user_id=owner_user_id)
    if not revoked:
        return None
    current = now or datetime.now(UTC)
    return McpTokenSummary(**token_summary_dict(record=revoked, now=current))

File: internal/mcptokens/service.go
package mcptokens

import (
	"context"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"strings"
	"time"

	"engram/internal/models"
	"engram/internal/repository"

	"github.com/google/uuid"
)

const (
	defaultTokenExpiryDays = 90
	defaultTokenListLimit  = 200
)

// Service orchestrates MCP token CRUD and token material handling.
type Service struct {
	db   repository.Queryer
	deps serviceDeps
}

type serviceDeps struct {
	nowUTC         func() time.Time
	newTokenID     func() uuid.UUID
	issueNewToken  func(input tokenIssueInput) (issuedToken, error)
	createMCPToken func(context.Context, repository.Queryer, repository.MCPTokenCreateInput) (*models.MCPTokenRecord, error)
	listMCPTokens  func(context.Context, repository.Queryer, repository.MCPTokenListInput) ([]models.MCPTokenRecord, error)
	revokeMCPToken func(context.Context, repository.Queryer, repository.MCPTokenRevokeInput) (*models.MCPTokenRecord, error)
}

type issuedToken struct {
	plaintext string
	hash      string
	hint      string
	expiresAt time.Time
}

// TokenHashInput captures token hashing materials.
type TokenHashInput struct {
	TokenID     uuid.UUID
	TokenSecret string
	Pepper      string
}

// TokenSecretVerificationInput captures secret verification materials.
type TokenSecretVerificationInput struct {
	TokenID      uuid.UUID
	TokenSecret  string
	Pepper       string
	ExpectedHash string
}

type tokenIssueInput struct {
	tokenID       uuid.UUID
	expiresInDays int
	pepper        string
}

// TokenListRequest captures list filters for token summaries.
type TokenListRequest struct {
	OwnerUserID uuid.UUID
	Limit       int
	Offset      int
}

func defaultServiceDeps() serviceDeps {
	return serviceDeps{
		nowUTC:        func() time.Time { return time.Now().UTC() },
		newTokenID:    uuid.New,
		issueNewToken: issueTokenWithDays,
		createMCPToken: func(
			ctx context.Context,
			db repository.Queryer,
			input repository.MCPTokenCreateInput,
		) (*models.MCPTokenRecord, error) {
			return repository.CreateMCPToken(ctx, db, input)
		},
		listMCPTokens: func(
			ctx context.Context,
			db repository.Queryer,
			input repository.MCPTokenListInput,
		) ([]models.MCPTokenRecord, error) {
			return repository.ListMCPTokens(ctx, db, input)
		},
		revokeMCPToken: func(
			ctx context.Context,
			db repository.Queryer,
			input repository.MCPTokenRevokeInput,
		) (*models.MCPTokenRecord, error) {
			return repository.RevokeMCPToken(ctx, db, input)
		},
	}
}

// NewService builds an MCP token service with repository defaults.
func NewService(db repository.Queryer) *Service {
	return &Service{db: db, deps: defaultServiceDeps()}
}

// NormalizeStringList trims, drops blanks, and deduplicates while preserving order.
func NormalizeStringList(values []string) []string {
	if len(values) == 0 {
		return []string{}
	}
	result := make([]string, 0, len(values))
	seen := make(map[string]struct{}, len(values))
	for _, raw := range values {
		value := strings.TrimSpace(raw)
		if value == "" {
			continue
		}
		if _, exists := seen[value]; exists {
			continue
		}
		seen[value] = struct{}{}
		result = append(result, value)
	}
	return result
}

// TokenHash computes the persisted hash for an MCP token secret.
func TokenHash(input TokenHashInput) string {
	payload := strings.ReplaceAll(input.TokenID.String(), "-", "") + ":" + input.TokenSecret
	mac := hmac.New(sha256.New, []byte(input.Pepper))
	_, _ = mac.Write([]byte(payload))
	return hex.EncodeToString(mac.Sum(nil))
}

// TokenSecretHint formats an operator-facing token hint.
func TokenSecretHint(tokenSecret string) string {
	if len(tokenSecret) <= 10 {
		return tokenSecret
	}
	return tokenSecret[:6] + "..." + tokenSecret[len(tokenSecret)-4:]
}

// BuildPlaintextToken composes the token string returned to the caller.
func BuildPlaintextToken(tokenID uuid.UUID, tokenSecret string) string {
	tokenIDHex := strings.ReplaceAll(tokenID.String(), "-", "")
	return models.MCPTokenPrefix + "_" + tokenIDHex + "_" + tokenSecret
}

// ParsePlaintextToken splits an MCP token into identifier and secret.
func ParsePlaintextToken(raw string) (uuid.UUID, string, error) {
	text := strings.TrimSpace(raw)
	prefix := models.MCPTokenPrefix + "_"
	if !strings.HasPrefix(text, prefix) {
		return uuid.UUID{}, "", errTokenPrefixMismatch
	}

	payload := strings.TrimPrefix(text, prefix)
	parts := strings.SplitN(payload, "_", 2)
	if len(parts) != 2 {
		return uuid.UUID{}, "", errTokenPayloadFormatMismatch
	}

	tokenIDHex := parts[0]
	tokenSecret := parts[1]
	if len(tokenIDHex) != 32 {
		return uuid.UUID{}, "", errTokenIdentifierLengthMismatch
	}
	tokenID, err := uuid.Parse(tokenIDHex)
	if err != nil {
		return uuid.UUID{}, "", errTokenIdentifierInvalid
	}
	if strings.TrimSpace(tokenSecret) == "" {
		return uuid.UUID{}, "", errTokenSecretMissing
	}
	return tokenID, tokenSecret, nil
}

// VerifyTokenSecret validates plaintext secret material against a persisted hash.
func VerifyTokenSecret(
	input TokenSecretVerificationInput,
) bool {
	candidateHash := TokenHash(
		TokenHashInput{
			TokenID:     input.TokenID,
			TokenSecret: input.TokenSecret,
			Pepper:      input.Pepper,
		},
	)
	return hmac.Equal([]byte(candidateHash), []byte(input.ExpectedHash))
}

// TokenIsActive evaluates token expiry and revocation state.
func TokenIsActive(record models.MCPTokenRecord, now time.Time) bool {
	if !now.IsZero() && record.ExpiresAt.IsZero() {
		return false
	}
	if record.RevokedAt != nil {
		return false
	}
	if now.IsZero() {
		now = time.Now().UTC()
	}
	return record.ExpiresAt.After(now)
}

// BuildAuthContext maps token storage record to authorization claims.
func BuildAuthContext(record models.MCPTokenRecord) models.MCPTokenAuthContext {
	return models.MCPTokenAuthContext{
		TokenID:           record.TokenID,
		OwnerUserID:       record.OwnerUserID,
		Scope:             record.Scope,
		AllowedTools:      NormalizeStringList(record.AllowedTools),
		AllowedProjectIDs: NormalizeStringList(record.AllowedProjectIDs),
		ExpiresAt:         record.ExpiresAt,
	}
}

// CreateTokenForOwner issues and persists a new MCP token for an owner.
func (service *Service) CreateTokenForOwner(
	ctx context.Context,
	ownerUserID uuid.UUID,
	payload models.MCPTokenCreateRequest,
	pepper string,
) (*models.MCPTokenCreateResponse, error) {
	scopeValue := resolveCreateScope(payload)
	parsedScope, err := models.ParseMCPTokenScope(scopeValue)
	if err != nil {
		return nil, err
	}
	tokenID := service.deps.newTokenID()
	issued, err := service.deps.issueNewToken(
		tokenIssueInput{
			tokenID:       tokenID,
			expiresInDays: resolveExpiresInDays(payload),
			pepper:        pepper,
		},
	)
	if err != nil {
		return nil, err
	}
	created, err := service.deps.createMCPToken(
		ctx,
		service.db,
		repository.MCPTokenCreateInput{
			TokenID:           &tokenID,
			OwnerUserID:       ownerUserID,
			Name:              strings.TrimSpace(payload.Name),
			Scope:             parsedScope,
			AllowedTools:      NormalizeStringList(payload.AllowedTools),
			AllowedProjectIDs: NormalizeStringList(payload.AllowedProjectIDs),
			TokenSecretHash:   issued.hash,
			TokenSecretHint:   issued.hint,
			ExpiresAt:         issued.expiresAt,
		},
	)
	if err != nil {
		return nil, err
	}
	return &models.MCPTokenCreateResponse{
		TokenID:           created.TokenID,
		Name:              created.Name,
		Scope:             created.Scope,
		AllowedTools:      created.AllowedTools,
		AllowedProjectIDs: created.AllowedProjectIDs,
		TokenSecretHint:   created.TokenSecretHint,
		Token:             issued.plaintext,
		ExpiresAt:         created.ExpiresAt,
		CreatedAt:         created.CreatedAt,
	}, nil
}

// ListTokenSummaries returns non-secret token metadata for an owner.
func (service *Service) ListTokenSummaries(
	ctx context.Context,
	request TokenListRequest,
) ([]models.MCPTokenSummary, error) {
	normalizedRequest := normalizeTokenListRequest(request)
	records, err := service.deps.listMCPTokens(
		ctx,
		service.db,
		repository.MCPTokenListInput{
			OwnerUserID: normalizedRequest.OwnerUserID,
			Limit:       normalizedRequest.Limit,
			Offset:      normalizedRequest.Offset,
		},
	)
	if err != nil {
		return nil, err
	}
	now := service.deps.nowUTC()
	summaries := make([]models.MCPTokenSummary, 0, len(records))
	for _, record := range records {
		summaries = append(summaries, buildTokenSummary(record, now))
	}
	return summaries, nil
}

// RevokeTokenForOwner revokes a token and returns the updated summary.
func (service *Service) RevokeTokenForOwner(
	ctx context.Context,
	tokenID uuid.UUID,
	ownerUserID uuid.UUID,
) (*models.MCPTokenSummary, error) {
	revoked, err := service.deps.revokeMCPToken(
		ctx,
		service.db,
		repository.MCPTokenRevokeInput{
			TokenID:     tokenID,
			OwnerUserID: ownerUserID,
		},
	)
	if err != nil {
		return nil, err
	}
	if revoked == nil {
		return nil, nil
	}
	summary := buildTokenSummary(*revoked, service.deps.nowUTC())
	return &summary, nil
}

func buildTokenSummary(record models.MCPTokenRecord, now time.Time) models.MCPTokenSummary {
	return models.MCPTokenSummary{
		TokenID:           record.TokenID,
		Name:              record.Name,
		Scope:             record.Scope,
		AllowedTools:      NormalizeStringList(record.AllowedTools),
		AllowedProjectIDs: NormalizeStringList(record.AllowedProjectIDs),
		TokenSecretHint:   record.TokenSecretHint,
		ExpiresAt:         record.ExpiresAt,
		LastUsedAt:        record.LastUsedAt,
		RevokedAt:         record.RevokedAt,
		CreatedAt:         record.CreatedAt,
		IsActive:          TokenIsActive(record, now),
	}
}

func resolveExpiresInDays(payload models.MCPTokenCreateRequest) int {
	if payload.ExpiresInDays <= 0 {
		return defaultTokenExpiryDays
	}
	return payload.ExpiresInDays
}

func normalizeTokenListRequest(request TokenListRequest) TokenListRequest {
	if request.Limit <= 0 {
		request.Limit = defaultTokenListLimit
	}
	if request.Offset < 0 {
		request.Offset = 0
	}
	return request
}

func resolveCreateScope(payload models.MCPTokenCreateRequest) string {
	trimmed := strings.TrimSpace(payload.Scope)
	if trimmed == "" {
		return string(models.MCPTokenScopeRead)
	}
	return trimmed
}

func issueTokenWithDays(input tokenIssueInput) (issuedToken, error) {
	expiresAt := time.Now().UTC().Add(time.Duration(input.expiresInDays) * 24 * time.Hour)
	return issueTokenWithExpiry(input, expiresAt)
}

func issueTokenWithExpiry(input tokenIssueInput, expiresAt time.Time) (issuedToken, error) {
	secretBytes := make([]byte, 32)
	if _, err := rand.Read(secretBytes); err != nil {
		return issuedToken{}, err
	}
	tokenSecret := base64.RawURLEncoding.EncodeToString(secretBytes)
	return issuedToken{
		plaintext: BuildPlaintextToken(input.tokenID, tokenSecret),
		hash: TokenHash(
			TokenHashInput{
				TokenID:     input.tokenID,
				TokenSecret: tokenSecret,
				Pepper:      input.pepper,
			},
		),
		hint:      TokenSecretHint(tokenSecret),
		expiresAt: expiresAt,
	}, nil
}
